src/main/java/org/nuxeo/wopi/rest/FilesResource.java

```diff
@@ -273,6 +276,66 @@ protected Object unlockOrRefresh(String fileId, String lock, boolean unlock) {
         }
     }
 
+    /**
+     * Implements the RenameFile operation.
+     * <p>
+     * See <a href="https://wopi.readthedocs.io/projects/wopirest/en/latest/files/RenameFile.html"></a>.
+     */
+    public Object renameFile(String fileId, String lock, String requestedName) {
+        NuxeoPrincipal principal = (NuxeoPrincipal) request.getUserPrincipal();
+        try (CloseableCoreSession session = createCoreSession(principal)) {
+            DocumentRef ref = new IdRef(fileId);
+            if (!session.exists(ref)) {
+                return Response.status(Response.Status.NOT_FOUND).build();
+            }
+
+            DocumentModel doc = session.getDocument(ref);
+            Blob blob = getMainBlob(doc);
+            if (blob == null) {
+                return Response.status(Response.Status.NOT_FOUND).build();
+            }
+
+            String nameJSON = "{ \"Name\": \"" + requestedName + "\" }";
+            if (!doc.isLocked()) {
+                if (!session.hasPermission(doc.getRef(), SecurityConstants.WRITE_PROPERTIES)) {
+                    // cannot rename blob
+                    return Response.status(Response.Status.CONFLICT).build();
+                }
+                blob.setFilename(requestedName);
+                BlobHolder bh = doc.getAdapter(BlobHolder.class);
+                bh.setBlob(blob);
+                doc.putContextData("source", "wopi");
+                doc = session.saveDocument(doc);
+                return Response.ok(nameJSON).build();
+            }
+
+            KeyValueService service = Framework.getService(KeyValueService.class);
+            KeyValueStore store = service.getKeyValueStore("wopi-locks");
+            String currentLock = store.getString(doc.getId());
+            if (currentLock == null) {
+                // locked by Nuxeo
+                return Response.status(Response.Status.CONFLICT).build();
```

Comment:
Is it possible to leverage `X-WOPI-LockFailureReason` for kind of lock ?

Reply:
Also, we might add the `X-WOPI-Lock` ?
>>>409 Conflict – Lock mismatch/locked by another interface; an X-WOPI-Lock response header containing the value of the current lock on the file must always be included when using this response code

Reply:
When the document is locked by Nuxeo, so not by WOPI, there is no lock value.
The `X-WOPI-LockFailureReason` is optional and it's only for logging purposes, and we don't even know if OO uses it. The point here is to have OO prompts the user that he can't rename the file.